Rework SemiPublic access and search

a2line opened the pull request:
Mostly a work on semipublic accessibility done by @hgouraud for Roglo database to fit with RGPD a few years ago, but that can be used on other databases.

hgouraud committed:
improved search:

search_by_name then search (not exact) for surname, then for first_name or public_name
in first input line, surnames is the "last word".
possibility of several first_names
mode buttons for first_name search (all, case, order, exact)

fmt pass

lex typo

# Conflicts:
#	hd/lang/lexicon.txt

hgouraud committed:
use referer after new passwd identification

hgouraud committed:
add SemiPublic to the access status and in gwc

remove duplicated lines to avoid build failure

hgouraud committed:
suppress trace message in util.ml and wiki.ml

hgouraud committed:
add is_related function testing relationship with SemiPublic  user

    # Conflicts:
    # hd/etc/menubar.txt

add semi_public and userip to conf; some fixup for semi_public

two commits to avoid build failure of first one.

hgouraud committed:
create caches folder if needed

hgouraud committed:
add p_auth_sp for SemiPublic;replace && by || in (not p_auth) || is_hide_names

lib/util.ml/mli remove chk default as gen_person_text option
to avoid build failure.

hgouraud committed:
manage Private situation

hgouraud committed:
Util.start_with is now available

hgouraud committed:
add comment for Mutil.tr

hgouraud committed:
read consent data from auth file

hgouraud committed:
add had_titles; remove chk in gen_person_txt

hgouraud committed:
suppress two unused variables

keep one empty line to avoid ocamlfmt error

hgouraud committed:
revise GWPARAM.p_auth

hgouraud committed:
add access status to history lines

modified to avoid ocamlfmt error

hgouraud committed:
x x -> individu/prénom/patronyme masqué

hgouraud committed:
try create cnt_dir with error report

hgouraud committed:
create caches folder in Request.init_etc

hgouraud committed:
add visibility.txt help file

hgouraud committed:
compute is_related only if e.fmode=on, otherwize reset the cache to 0

hgouraud committed:
remove trace in searchName.ml

hgouraud committed:
provide fn sn to WLperson if no name

hgouraud committed:
rework linked_pages_text

hgouraud committed:
fix update_nldb fot name and text mixup

hgouraud committed:
sort old wizards list

hgouraud committed:
add -bd to update_nldb, -debug mode

hgouraud committed:
correct MOD_WIZNOTES call; correct wizard name in wiznotes list

hgouraud committed:
make m=H available

hgouraud committed:
add user.index in template

a2line committed:
small template fix for Roglo

a2line committed:
more responsive welcome

hgouraud committed:
reduce forum spacing; set wiki=on

hgouraud committed:
suppress bg-white in welcome, menubar updind/updfam

hgouraud committed:
add 3 pointers to legal data in copyr.txt

hgouraud committed:
bring lang selection to top of welcome page for Roglo

hgouraud committed:
remove unwanted </li> in middle of linked_page.NOTE

* remove dots (.) in notes template after linked_page.NOTE
* rework linked_page_text in perso.ml
* update failing wiki_test

hgouraud committed:
add header message in forum

a2line committed:
fix broken updfam in templm

broken since change introduced in 0e94f9ab6b8f717a238f1ac42450a68ce9dd5f9e

a2line committed:
mostly template fixes when using background image

* restore button on horizontal trees
* update margin on forms, more homogen insert functions
* load forgoten regex.min.js for books
* rework titles on short trees

hd/etc/modules/arbre_vertical.txt use e_v and not e.v
to avoid test/run_gw_test.sh failure on arbres test with p_mod=a1

hgouraud committed:
for roglo, link to friend/wizard base on welcome page

hgouraud committed:
add semi_public=yes in .gwf

this restores standard behaviour if semi_publis absent or =no
some optimization in p_auth
remove semi_public buttons in menubar and updind.txt
test existence of family_mode cache file before delete

hgouraud committed:
add is_person in history.ml

hgouraud committed:
prevent wizard actions on roglo

hgouraud committed:
use !gwparam.p_auth_sp instead of not p_auth for occ

hgouraud committed:
capitalize 0 length string! better title for m=D

a2line committed:
red button to delete

hgouraud committed:
replace “in (place)” by a dash for Roglo

hgouraud committed:
translations in visibility.txt: suppress reset family mode in lexicon

hgouraud committed:
use [send::] (add, modify, update, delete, merge)  to handle correctly german and nl

a2line committed:
missing etc_prefix for CGI

fixes #2070

hgouraud committed:
rework copyr (test before <div>) and visibility

hgouraud committed:
three inputs search

* adjust home and auth section
* use . to separate fn from sn in first search line
* split result list in two parts exact
* use "/" as fn/sn/oc delimiter, better search for spouses
* search for public_name surname
* adjust welcome input area tooltip
* show directly person as often as possible
* keep input as typed, remove occ, sort by b_date, use public_name
* exact and all by default for first_name search
* homogeneize fn/sn/oc behaviour for 1st input
* search first_names with AdvancedSearch, provides pl1, pl2
* first_name search (wip)

hgouraud committed:
dont create init files (cache, etc, cnt, ...) if base.gwb does not exist

a2line committed:
Better encode and escaping HTML functions for books

* Handles type conversions for encode/escape functions by using intermediate
  variables instead of direct coercion that was causing various errors)
* Apply `move_particle` function only to surnames and domains, not to all books
* Filter out dummy placeholder “?” entries from first name and surname lists
* `Util.uri_encode` replaces `Mutil.encode`, %uri_encode.variable; syntax now
  match the behaviour of escape function making template usage more consistent.
  - Depreciate confusing `html_encode` alias for escape.
  - Depreciate `safe`(_html) unused and not needed on books.
* rewrite substring function to support `%escape.substring;`/`%encode.substring;`
  the breadcrumbs trail now uses only `%foreach;substring.s.n;` (n limited to 10).
* Add a JS function to shorten sup-char values when too many entries left.
* Rewrite `%subtitle;` variable to escape its third value adding a blue highlight.
* Resolves several HTML validation errors: Fix unclosed div elements and mismatched
  tags in list items, properly structure forms elements to maintain valid HTML and
  ensure correct nesting of conditional sections.
* Add Bootstrap tooltips for better accessibility, remove default HTML=true
  option to prevent escaping failure, it can be added where needed.

This should definitively fix #2054.

hgouraud committed:
update carrousel even if no portrait

Commit:
fix ancestor v1

fixes #2082

a2line committed:
detail main search tooltip, add examples

more responsive rendering
better manage b.propose_titles parameter

a2line committed:
move new note button, missing image map translation

* Move add new note button to welcome to lighten menubar
* Add the missing translation title for image map initialization

hgouraud committed:
do not propose person (in specify) if p_auth=false

hgouraud committed:
color perso-tools icon according to visibility

hgouraud committed:
alt_base for everybody; show MISC_NOTES; roglo_encyclopedia only for Roglo

hgouraud committed:
improve visibility.txt

hgouraud committed:
rework options for search first_name

commit updated to avoid build failure, and pass ocamlformat
by dragging some changes from next commit.

michelmno commented:
I found three problems with this branch:

1- dune @fmt has identified few files to be reformatted  as per my local commit https://github.com/michelmno/geneweb/commit/2d1f462c06d484b07ecfef6dd83e11b5ac56ac42
   Update 20250313:
   Oups, this first problem was related to a2line/roglo-7 AND NOT hgouraud/roglo-7 that is source of the PR, my mistake.
   But two following problems remain.

2- Furthermore if I am trying to run test/run_gw_test.sh on my local database, then many warnings about cnt creation are reported as per gwd log extract; while cnt already exist as per stat output.
```
$head -n15 /tmp/gwd_cgi_tst_error.log
2025-03-12 19:38:15 (1639469) gwtst?b=basefusionnee&w=azerty:dcba&
  From: 127.0.0.1
  Agent: curl/7.81.0
[2025-03-12 19:38:15]: WARNING Failure when creating cnt
  #accesses 406,878 (#welcome 15,618) since 17/01/2007
2025-03-12 19:38:15 (1639477) gwtst?b=basefusionnee&w=azerty:dcba&m=S&n=anne+normand&p=
  From: 127.0.0.1
  Agent: curl/7.81.0
[2025-03-12 19:38:15]: WARNING Failure when creating cnt
  #accesses 406,879 (#welcome 15,618) since 17/01/2007
2025-03-12 19:38:16 (1639485) gwtst?b=basefusionnee&w=azerty:dcba&p=anne&n=normand&oc=5
  From: 127.0.0.1
  Agent: curl/7.81.0
[2025-03-12 19:38:16]: WARNING Failure when creating cnt
  #accesses 406,880 (#welcome 15,618) since 17/01/2007
```
```
$stat /var/lib/geneweb/cnt
  Fichier : /var/lib/geneweb/cnt
   Taille : 12288     	Blocs : 24         Blocs d'E/S : 4096   répertoire
Périphérique : 10302h/66306d	Inœud : 17599844    Liens : 2
Accès : (0755/drwxr-xr-x)  UID : (   33/www-data)   GID : (   33/www-data)
Accès : 2025-03-12 16:56:01.953368463 +0100
Modif. : 2025-03-12 14:58:44.114486412 +0100
Changt : 2025-03-12 14:58:44.114486412 +0100
  Créé : 2024-07-08 06:59:12.812817660 +0200
```

3- and ultimately trying to build each commit of the branch, the second one is failing

```
$git rebase -x ~/work/geneweb/local/checkgwuodir 0861ab4b1 
...
Repository michelmno/geneweb. Branch HEAD. Last commit 5fd67e469 message:
  add SemiPublic to the access status and in gwc
...
Building executables
(cd _build/default && /home/michel/.opam/sandbox4.14.2/bin/not-ocamlfind preprocess -package camlp5.extend,camlp5.quotations,camlp5.pr_o -syntax camlp5o bin/ged2gwb/ged2gwb.ml) > _build/default/bin/ged2gwb/ged2gwb.pp.ml
ocamlfind: [WARNING] Package `camlp5.extend': camlp5.extend SHOULD NOT be used with syntax camlp5o
Restoring original patched files… Done.   
File "lib/GWPARAM.mli", lines 51-57, characters 0-5:
Error (warning 32 [unused-value-declaration]): unused value output_error.
File "lib/GWPARAM.mli", line 62, characters 0-76:
Error (warning 32 [unused-value-declaration]): unused value is_semi_public.
File "lib/GWPARAM.mli", line 65, characters 0-68:
Error (warning 32 [unused-value-declaration]): unused value p_auth.
File "lib/GWPARAM.mli", line 68, characters 0-49:
Error (warning 32 [unused-value-declaration]): unused value syslog.
File "lib/GWPARAM.mli", lines 71-72, characters 0-67:
Error (warning 32 [unused-value-declaration]): unused value wrap_output.
make: *** [Makefile:259 : ci] Erreur 1             
warning: échec d'exécution : /home/michel/work/geneweb/local/checkgwuodir
Vous pouvez corriger le problème, puis lancer
git rebase --continue
```

a2line commented:
If it's passing CI there are no fmt errors.

michelmno commented:
> If it's passing CI there are no fmt errors.

my mistake for the first problem as per updated comment.
But two other problems remain.

hgouraud committed:
use some.ml functions to print list of first_names

michelmno commented:
@hgouraud  I worked on your roglo-7 branch head cid  739de7d2a, on the 66 current commits  of the PR I identified 7 commits  that needed some changes to avoid build and test errors.
Those changes are part  of my own roglo-7 branch https://github.com/michelmno/geneweb/commits/roglo-7/ 
I would suggest to continue your work on top of it as would solve the third problem referenced in my initial comment https://github.com/geneweb/geneweb/pull/2055#issuecomment-2718868065

The 7 commits are:
* 6cf325762 rework options for search first_name
    need to drag some lines from next commit 249a93eec
    to avoid build error.
* 9fdf07db1 mostly template fixes when using background image
    one line change to avoid test/run_gw_test.sh detected error
* f8c2dfebe add access status to history lines
    ocamlfmt updated
* 4613a2fba suppress two unused variables
    ocamlfmt updated
* 8a1560cba add p_auth_sp for SemiPublic;replace && by || in (not p_auth) || is_hide_names
    complete patch removing chk default as gen_person_text option in lib/util.ml*
* c5d1115ae add semi_public and userip to conf; some fixup for semi_public
    squashed with 343314de6 to avoid build error of last one.
  343314de6 add is_related function testing relationship with SemiPublic  user
* 02d0e124c add SemiPublic to the access status and in gwc
    in lib/GWPARAM.mli remove some duplicated lines done only on next commit

to be compared to those in my branch
* afe4dff56 rework options for search first_name
* 6124eae4d mostly template fixes when using background image
* c59c75af8  add access status to history lines
* 251bb14c1 suppress two unused variables
* fe264b747  add p_auth_sp for SemiPublic;replace && by || in (not p_auth) || is_hide_names
* fbfee01f8    add is_related function testing relationship with SemiPublic  user
* 8a5b30262 add SemiPublic to the access status and in gwc

hgouraud committed:
fix performance issue on bold/italic

This commit fixes b7b62e111e4e49af5124c2f9ecceff2ab771d0f1 that
introduced tremendous latency opening long notes.

hgouraud committed:
in hutil.ml, set container-fluid if b.wide or e.wide=on"

hgouraud committed:
dont init_etc if base.gwb does not exist

and changes to avoid ocamlfmt error.

hgouraud committed:
sort by date surname lists

michelmno commented on lib/util.ml:
```diff
@@ -1753,7 +1753,36 @@ let hexa_string s =
+    pl
+
+let print_alphab_list conf base crit print_elem liste =
+  let print_sub_list base len sub_l print_e index =
```
build error for the same reason: base not used because line 1774 is a comment line.

michelmno commented:
@hgouraud, to continue my comments on https://github.com/geneweb/geneweb/pull/2055#issuecomment-2726215416
I rebased my own branch roglo-7 with same modified commits plus two.
This is required to have them to properly build and ease potential git bissect work on upstream/master.

The modified commits in my branch are:

* 0b3837c7c dont init_etc if base.gwb does not exist
* 6347be735 use some.ml functions to print list of first_names (wip)
* fbd2aba09 rework options for search first_name
* 100346a27 mostly template fixes when using background image
* 16e1c5e86 add access status to history lines
* 332622556 suppress two unused variables
* d65d3bc70 add p_auth_sp for SemiPublic;replace && by || in (not p_auth) || is_hide_names
* 3907c244c add is_related function testing relationship with SemiPublic  user
* 2303628cd add SemiPublic to the access status and in gwc

To confirm no functional changes, able to run a git diff between two branches like
`$git diff hgouraud/roglo-7 michelmno/roglo-7`

hgouraud commented:
1/ I see a few diff (possibly irrelevant) between hgouraud/roglo-7 and michelmno/roglo-7 
2/ comment faire ensuite?

michelmno commented:
I identified the diff as per 
[diff_hgouraud_michelmno.txt](https://github.com/user-attachments/files/19483954/diff_hgouraud_michelmno.txt)
my explanation of those diff:
* arbre_vertical.txt as per commit
  100346a27 mostly template fixes when using background image
  completion to avoid test/run_gw_test.sh failure
* lib/some.ml as per commit
  6347be735 use some.ml functions to print list of first_names (wip)
  blank lines added by ocamlfmt
* lib/util.mli as per commit
  d65d3bc70 add p_auth_sp for SemiPublic;replace && by || in (not p_auth) || is_hide_names
  modified to avoid build error, but also associated comment.

What I would suggest locally in your PC: keep your roglo-7 branch as backup, replace by  own roglo-7 branch before to force it as your 'origin' repo:
```
$git branch -m roglo-7 roglo7-backup
$git checkout -b roglo-7 michelmno/roglo-7
$git git push -f --set-upstream origin
```
Assumption: you are working as 'origin' repo and mine as 'michelmno':
```
$git remote -v |grep -E 'michelmno|origin'
michelmno	https://github.com/michelmno/geneweb (fetch)
michelmno	https://github.com/michelmno/geneweb (push)
origin	git@github.com:hgouraud/geneweb.git (fetch)
origin	git@github.com:hgouraud/geneweb.git (push)
```

hgouraud commented:
ok, je comprend. Ferai cet après midi!

Done

hgouraud committed:
do not print error_list if empty. Correct bad % in url in welcome.txt

hgouraud committed:
fix undefined variable in anctree.txt, arbre_8gen.txt and fratrie.txt

hgouraud committed:
correct boolean equation for 'file' in home.txt

hgouraud committed:
small update in run_gw_test

michelmno commented:
> 2- Furthermore if I am trying to run test/run_gw_test.sh on my local database, then many warnings about cnt creation are reported as per gwd log extract; while cnt already exist as per stat output.

I am suggesting a patch to avoid this failure, as my commit https://github.com/michelmno/geneweb/commit/7e065f41a8dc4b8ed843188d0e4c90e1415b8dce

Halbaroth commented on lib/wiki.mli:
```diff
@@ -35,14 +35,15 @@ open Config
 
 val make_edit_button :
-  Config.config -> string -> ?cnt:int option -> unit -> string
+  Config.config -> string -> string -> ?cnt:int option -> unit -> string
```
I suggest to use named argument when the type is not sufficient to understand the purpose of an argument. 
```suggestion
  Config.config -> string -> mode:string -> ?cnt:int option -> unit -> string
```

Halbaroth replied:
```suggestion
  Config.config -> string -> mode:string -> ?cnt:int option -> unit -> string
```

Halbaroth commented on plugins/forum/forumDisplay.ml:
```diff
@@ -377,7 +378,7 @@ let print_add_ok conf base =
       m_access = "";
       m_subject = subject;
-      m_wiki = "";
+      m_wiki = "on";
```
Does `m_wiki` take its values from a finite set? If so, it is better to use an enum type instead of a string.

Halbaroth commented on lib/util/utf8.ml:
```diff
@@ -78,18 +78,20 @@ let skip_html_tags s =
+      else `Self
+    in
+    let tail = cmap_utf_8 cmap tail in
+    head ^ tail
```
Can you explain the purpose of this change? The old code produces the same result on the empty string. I tested.

Halbaroth commented on lib/util/mutil.ml:
```diff
@@ -244,6 +244,7 @@ let surnames_pieces surname =
   in
   loop 0 0 0
 
+(** in string s, replace c1 by c2 *)
```
This function is already documented in the mli as follows: 
```ocaml
(** [tr c1 c2 str]
    Return a new string which is the same as [str] with all occurences of [c1]
    replaced by [c2]. If [str] does not contain [c1] [str] is returned untouched.
 *)
```

Halbaroth commented on lib/util.mli:
```diff
@@ -369,10 +367,8 @@ val string_of_witness_kind_raw : witness_kind -> Adef.safe_string
-  config -> base -> string * string * int -> bool * string * string
-
+val person_exists : config -> base -> string * string * int -> bool
+val mark_if_not_public : config -> base -> string * string * int -> bool
```
Please add documentation.

Halbaroth commented on lib/config.mli:
```diff
@@ -56,10 +56,12 @@ type config = {
   user : string;
   username : string;
   userkey : string;
+  userip : iper option;
```
Documentation again :)
The name `userip` is misleading. I believe that `user_iper` is better.

Halbaroth commented on bin/gwc/gwc.ml:
```diff
@@ -164,7 +164,10 @@ let errmsg =
+  (try
+     if Sys.is_directory !Gwcomp.rgpd_files then Gwcomp.rgpd := true
+     else Gwcomp.rgpd := false
+   with Sys_error _ -> Gwcomp.rgpd := false);
```
This code silently disables the RGPD feature if the user does not use it correctly, which could be really confusing. Instead, we should display a warning message explaining why the feature cannot be enabled.

Halbaroth commented on bin/gwc/gwc.ml:
```diff
@@ -131,18 +130,19 @@ let speclist =
-    ("-q", Arg.Clear Mutil.verbose, " Quiet");
+    ("-q", Arg.Clear Gwcomp.verbose, " Quiet");
     ("-reorg", Arg.Set Geneweb.GWPARAM.reorg, " Mode reorg");
+    ("-rgpd", Arg.String (fun s -> Gwcomp.rgpd_files := s), "<file> Rgpd files");
```
It seems that `-rgpd` expects a directory path, not a file path.
```suggestion
    ("-rgpd", Arg.String (fun s -> Gwcomp.rgpd_dir := s), "<dir> Rgpd directory");
```

Halbaroth commented on bin/update_nldb/update_nldb.ml:
```diff
@@ -261,18 +268,18 @@ let compute base bdir =
   let () = load_strings_array base in
   let () = load_unions_array base in
-  try compute base !fname
+  try compute base (Filename.concat (Secure.base_dir ()) !fname)
```
You already compute this concatenation of paths above, let's reuse it!
Or you can modify `compute` because the name of the database can be obtained with a function `Gwd_driver`.

Halbaroth commented on bin/update_nldb/update_nldb.ml:
```diff
@@ -22,15 +31,8 @@ let notes_links s =
-      | NotesLinks.WLperson (j, key, name, text) ->
+      | NotesLinks.WLperson (j, key, _name, text) ->
           let list_ind =
-            let text =
-              match (name, text) with
-              | Some name, None -> Some name
-              | Some name, Some text when name = "" -> Some text
-              | Some name, Some text -> Some (name ^ ";" ^ text)
-              | _, _ -> None
-            in
```
I don't understand why this code have been erased.

Halbaroth commented on lib/perso.ml:
```diff
@@ -1104,23 +1104,23 @@ let linked_page_text conf base p s key (str : Adef.safe_string) (pg, (_, il)) :
-                      (Adef.safe "", Util.safe_html v, Adef.safe "")
+                      (* FIXME Util.safe_html was introducing unwanted </li>  *)
+                      (a, b, c)
+                    with Not_found -> ("", v, "")
```
This bug should be fix in this PR or you might open an issue to explain the issue and add a link to this issue in the comment.

Halbaroth commented on lib/perso.ml:
```diff
@@ -3684,23 +3709,24 @@ and eval_str_person_field conf base env ((p, p_auth) as ep) = function
+      else str_val (Util.private_txt conf "p")
   | "first_name_key" ->
-      if is_hide_names conf p && not p_auth then null_val
+      if (not p_auth) || is_hide_names conf p then null_val
```
You use this condition several times. Creating a dedicated function is appropriate and its name would be a good way to document this piece of code.

Halbaroth commented on lib/perso.ml:
```diff
@@ -3565,11 +3586,16 @@ and eval_bool_person_field conf base env (p, p_auth) = function
       | _ -> raise Not_found)
   | "is_female" -> get_sex p = Female
   | "is_invisible" ->
+      (* to visitors *)
```
I don't understand this comment, can you explain yourself?

Halbaroth commented on lib/image.ml:
```diff
@@ -271,28 +271,30 @@ let rename_portrait conf base p (nfn, nsn, noc) =
+      ();
+      (* in all cases, rename carrousel *)
+      let new_c_f =
         String.concat Filename.dir_sep [ carrousel_folder conf; new_s ]
```
It is not a part of the PR, but `Filename.concat` is more appropriate than `String.concat Filename.dir_sep`.

Halbaroth commented on lib/notesDisplay.ml:
```diff
@@ -51,7 +51,7 @@ let print_whole_notes conf base fnotes (title : Adef.safe_string) s ho =
   in
   let modbtn =
-    if conf.wizard then Wiki.make_edit_button conf fnotes () else ""
+    if conf.wizard then Wiki.make_edit_button conf "NOTES" fnotes () else ""
```
```suggestion
    if conf.wizard then Wiki.make_edit_button conf ~mode:"NOTES" fnotes () else ""
```

Halbaroth commented on bin/gwd/gwd.ml:
```diff
@@ -1169,15 +1169,21 @@ let make_conf from_addr request script_name env =
      api_mode = false;
      manitou = manitou;
-     supervisor = supervisor; wizard = ar.ar_wizard && not wizard_just_friend;
+     supervisor = supervisor;
```
You can write `supervisor;` instead of `supervisor = supervisor;`

hgouraud committed:
fmt pass and various fix following review

Halbaroth commented on bin/gwd/gwd.ml:
```diff
@@ -1802,7 +1808,10 @@ let geneweb_server () =
+        try File.create_dir ~parent:true ~required_perm:0o755 !GWPARAM.cnt_dir
+        with Sys_error e ->
+          GwdLog.syslog `LOG_CRIT (Format.sprintf "failure creating %s (%s)\n"
+            !GWPARAM.cnt_dir e);
```
I introduced this code during the last sprint. It was a mistake to enforce the permissions because user can change this permissions or older version can use different permissions. 

For the server, it is a fatal error to fail to create `cnt_dir`. You should stop it instead of logging a message. I think that the previous code was better because the OCaml runtime will provide a backtrace for the uncaught exception and the server will stop because of it.

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -2,6 +2,7 @@
 
 open Def
 open Gwdb
+open Geneweb
```
Please do not open modules globally in another module. Namespaces are helpful to localize definitions and to avoid undesired shadowing of values. I know that the codebase is full of this bad habit but we should not introduce more opens. Instead, you can qualify values with `Geneweb.`.

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -10,6 +11,27 @@ let magic_gwo = "GnWo000o"
 (* si la clé est incomplète, on l'enregistre tout de même.  *)
 let create_all_keys = ref false
 
+(** Line counter while reading .gw file *)
```
Nit: the odoc comment `(** ... *)` does not work in a ml module and I think that odoc will emit a warning if you use it in such modules. If the value is exposed in the mli, you move the documentation in the mli. Otherwise, you can replace it with a common comment `(* .. *)`.

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -10,6 +11,27 @@ let magic_gwo = "GnWo000o"
+let line_cnt = ref 0
+
+(** Do not raise exception if syntax error occured.
+    Instead print error information on stdout *)
```
Same

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -10,6 +11,27 @@ let magic_gwo = "GnWo000o"
+    Instead print error information on stdout *)
+let no_fail = ref false
+
+(** Save path to the images *)
```
Same

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -10,6 +11,27 @@ let magic_gwo = "GnWo000o"
+(** Save path to the images *)
+let no_picture = ref false
+
+(** Fonctionnement RGPD *)
```
Same and I don't understand this comment.

Halbaroth commented on bin/gwc/gwcomp.mli:
```diff
@@ -3,6 +3,13 @@
+
+val rgpd : bool ref
+val verbose_friends : bool ref
+val semi_pub_cnt : int ref
```
Please document this values.

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -523,6 +535,128 @@ let get_access l =
+            with End_of_file ->
+              close_in ic;
+              Some "")
+        | None -> Some "")
```
I suggest to replace this snippet with the following code:
```ocaml 
        try
          Secure.with_open_in_text gwf_file (fun ic ->
            let rec loop () =
              match input_line ic with
              | exception End_of_file -> Some ""
              | line when Util.start_with line 0 "friend_passwd_file" -> (
                  match extract_value line with
                  | exception Not_found -> Some ""
                  | passwd -> Some passwd)
              | line -> loop ()
            in
            loop ())
        with Sys_error _ -> Some "")
```
where `extract_value` is defined as: 
```ocaml  
(* Assuming that the string [s] is of the form [key=value], this
   function extracts the value.

   @raise Not_found if [s] does not contain exactly one delimiter '='. *)
let extract_value s =
  let len = String.length s in
  let rec loop equal_pos i =
    if i = len then equal_pos
    else
      let c = s.[i] in
      match equal_pos with
      | Some _ when c = '=' ->
          (* The line contains several delimiters. *)
          None
      | None when c = '=' ->
          (* We found the first delimiter. *)
          loop (Some i) (i + 1)
      | _ -> loop equal_pos (i + 1)
  in
  match loop None 0 with
  | Some i -> String.sub s (i + 1) (len - i - 1)
  | None -> raise Not_found
```

Now, let me explain the changes: 
1. The original code includes the following construct: 
```ocaml 
match try Some (Secure.open_in gwf_file) with Sys_error _ -> None with
        | Some ic -> (
```
This pattern is recurrent in the codebase and is both inefficient and less readable:
- It allocates an option value and destroy it just after. I am not sure that the compiler can optimize this case. 
- It is more verbose that the modern and idiomatic approach:
```ocaml 
match Secure.open_in friend_passwd_file with 
| exception Sys_error _ -> ...
| ic -> ... 
```
2. Manually closing channels within exception handlers is tricky and error-prone. Please do not reintroduce this pattern. I added a function `with_open_in_text`  in the `Secure` module to handle this situation. The code is safer and easier to maintain.
3. The original code used:
```ocaml 
                  let parts = String.split_on_char '=' line in
                  close_in ic;
                  if List.length parts = 2 then Some (List.nth parts 1)
                  else Some "")
```
which extracts the password in the current line. First, `parts` will construct a list even if the line has not the right format. The `List.length` will iterate on all the list to compute its length and compare it with two, the function `List.compare_length_with` is more appropriate for this task. The function `extract_value` performs only two allocations.

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -523,6 +535,128 @@ let get_access l =
+            in
+            match
+              try Some (Secure.open_in friend_passwd_file)
+              with Sys_error _ -> None
```
Please use `Secure.with_open_in_text`.

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -523,6 +535,128 @@ let get_access l =
+            | Some ic -> (
+                let acc = ref [] in
+                try
+                  let rec loop () =
```
You can pass `acc` as an argument of `loop`. References are a bit slower and less readable.

Halbaroth commented on bin/gwc/gwcomp.ml:
```diff
@@ -523,6 +535,128 @@ let get_access l =
+  in
+  let is_consent =
+    List.mem (Format.sprintf "%s.%d+%s" fns oc sns) consent_list
+  in
```
Comparing strings is very slow. You should use a hash table for `consent_list`.

Halbaroth commented on bin/update_nldb/update_nldb.ml:
```diff
@@ -2,9 +2,18 @@ open Geneweb
 
+let debug = ref false
 let fname = ref ""
 let errmsg = "usage: " ^ Sys.argv.(0) ^ " [options] <file_name>"
```
Nit: it is easier to read: 
```ocaml 
Format.sprintf "usage: %s [options] <file_name>" Sys.argv.(0) 
```

Halbaroth commented on bin/update_nldb/update_nldb.ml:
```diff
@@ -2,9 +2,18 @@ open Geneweb
+    ( "-bd",
+      Arg.String Secure.set_base_dir,
+      "<DIR> Specify where the “bases” directory with databases is installed \
+       (default if empty is “.”)." );
```
I am not sure that `“` is displayed correctly with all the terminals. Using simple quote is probably more portable.

Halbaroth commented on lib/GWPARAM.ml:
```diff
@@ -137,145 +140,359 @@ module Legacy = struct
-    content ();
-    Output.print_sstring conf {|</body></html>|}
-end
+    match (List.assoc_opt "fmode" conf.Config.env :> string option) with
```
```ocaml 
match (List.assoc "fmode" conf.Config.env :> string) with 
| exception Not_found -> ... 
| "on" -> ... 
....
```

Halbaroth commented on lib/GWPARAM.ml:
```diff
@@ -137,145 +140,359 @@ module Legacy = struct
+              match conf.Config.userip with
+              | Some ip ->
+                  let family = [ ip ] in
+                  let max =
+                    try List.assoc "is_semi_public_max" conf.Config.base_env
+                    with Not_found -> "2" |> String.trim
+                    (* limit search to n generations *)
+                  in
+                  let max = if max = "" then 2 else int_of_string max in
```
```ocaml 
let max = 
   try 
      List.assoc "is_semi_public_max" conf.Config.base_env |> String.trim |> int_of_string 
   with 
   | Not_found -> 2
   | Failure "int_of_string" -> (* handle this error here *)
```

Halbaroth commented on lib/GWPARAM.ml:
```diff
@@ -137,145 +140,359 @@ module Legacy = struct
+                  in
+                  let max = if max = "" then 2 else int_of_string max in
+                  let family = ancestors conf base (max + 1) family ip in
+                  (* siblings
+                     let family =
+                           (match Gwdb.get_parents (Gwdb.poi base ip) with
+                           | Some ifam -> Gwdb.get_children (Gwdb.foi base ifam) |> Array.to_list
+                           | None -> [])
+                           @ family
+                     in *)
```
Should we keep this comment?

Halbaroth commented on lib/GWPARAM.ml:
```diff
@@ -137,145 +140,359 @@ module Legacy = struct
+              | Some ip ->
+                  let family = [ ip ] in
+                  let max =
+                    try List.assoc "is_semi_public_max" conf.Config.base_env
```
The variable name `is_semi_public_max` is no self-documenting.

Halbaroth commented on lib/GWPARAM.mli:
```diff
@@ -116,18 +94,32 @@ module Legacy : sig
+  unit)
+  ref
+(** If [?content] is not set, sends page content from [/etc/<status-code>-<lang>.html].
       If the current lang is not available, use `en` *)
```
Please do not move code that does not change. The diff is harder to read and the git history will be less readable. 

Do not use a reference here.

Halbaroth commented on lib/GWPARAM.mli:
```diff
@@ -116,18 +94,32 @@ module Legacy : sig
+
+val p_auth : (Config.config -> Gwdb.base -> Gwdb.person -> bool) ref
+(** Calculate the access rights to the person's information in
       according to his age.
```
There is no reason to use reference on functions here. Calling a function through a reference is slower.

Halbaroth commented on lib/GWPARAM.mli:
```diff
@@ -146,9 +138,12 @@ module Legacy : sig
+
+val wrap_output :
+  (Config.config -> Adef.safe_string -> (unit -> unit) -> unit) ref
+(** Display in a very basic HTML doc, with no CSS or JavaScript. *)
```
Same here, no reason to use reference.

Halbaroth commented on lib/GWPARAM.mli:
```diff
@@ -116,18 +94,32 @@ module Legacy : sig
+val is_semi_public : (Gwdb.person -> bool) ref
+(** determines if the person has SemiPublic status   *)
+
+val split_key : string -> string * string * string
```
Please add documention.

Halbaroth commented on lib/config.ml:
```diff
@@ -51,10 +51,12 @@ type config = {
   user : string;
   username : string;
   userkey : string;
+  userip : iper option;
```
This name is not self-documenting and is misleading. I think that `user_iper` is clearer.

Halbaroth commented on lib/hutil.ml:
```diff
@@ -113,12 +113,25 @@ let header_without_http_nor_home conf title =
+    (try List.assoc "wide" conf.env = Adef.encoded "on"
+     with Not_found -> false)
+    || try List.assoc "wide" conf.base_env = "on" with Not_found -> false
+  in
```
```ocaml 
let header_without_title conf =
  let is_fluid = is_fluid conf in 
  ...
```
where `is_fluid` is defined as:
```ocaml
  let is_fluid conf =
    let is_wide env = try List.assoc "wide" env = Adef.encoded "on" with Not_found -> false in 
    is_wide conf.env || is_wide conf.base_env 
```

Halbaroth commented on lib/hutil.ml:
```diff
@@ -113,12 +113,25 @@ let header_without_http_nor_home conf title =
+    || (try List.assoc "wide" conf.env = Adef.encoded "on"
+        with Not_found -> false)
+    || try List.assoc "wide" conf.base_env = "on" with Not_found -> false
+  in
```
Reuse the function `is_fluid` is suggested above.

Halbaroth commented on lib/hutil.ml:
```diff
@@ -134,10 +147,17 @@ let header_fluid conf title = header_with_title ~fluid:true conf title
+    (try List.assoc "wide" conf.env = Adef.encoded "on"
+     with Not_found -> false)
+    || try List.assoc "wide" conf.base_env = "on" with Not_found -> false
+  in
```
again, reuse `is_fluid`.

Halbaroth commented on lib/image.ml:
```diff
@@ -271,28 +271,30 @@ let rename_portrait conf base p (nfn, nsn, noc) =
+          !GWPARAM.syslog `LOG_ERR
+            (Format.sprintf
+               "Error renaming old portrait: old_path=%s new_path=%s : %s" old_f
+               new_f e))
```
```suggestion
     let fl = old_s_f ^ old_ext in
      if Sys.file_exists fl then
        try Sys.rename fl (new_s_f ^ old_ext)
        with Sys_error e ->
          !GWPARAM.syslog `LOG_ERR
            (Format.sprintf
               "Error renaming old portrait: old_path=%s new_path=%s : %s" old_f
               new_f e))
```

Halbaroth commented on lib/notesLinks.ml:
```diff
@@ -132,7 +132,8 @@ let misc_notes_link s i =
                 let oc = try int_of_string oc with Failure _ -> 0 in
-                if oc1 = -1 then (fn, sn, oc, Some name) else (fn, sn, oc1, None)
+                if oc1 = -1 then (fn, sn, oc, Some name)
+                else (fn, sn, oc1, Some (fn ^ " " ^ sn))
```
Ok, it seems that you change the meaning of the last field of note links. We should document it.

Halbaroth commented on lib/perso.ml:
```diff
@@ -1104,23 +1104,23 @@ let linked_page_text conf base p s key (str : Adef.safe_string) (pg, (_, il)) :
-                        : Adef.escaped_string
-                        :> Adef.safe_string)
-                  ^^^ {|">|} ^<^ b ^^^ {|</a>|} ^<^ c
+                  Printf.sprintf "%s<a href=\"%sm=NOTES;f=%s#p_%d\">%s</a>%s" a
+                    (commd conf :> string)
+                    (Mutil.encode pg :> string)
+                    text.Def.NLDB.lnPos b c
+                  |> Util.safe_html
```
It is far better :)

Halbaroth commented on lib/perso.ml:
```diff
@@ -1329,17 +1329,29 @@ let make_efam conf base ip ifam =
-      match p with
-      | None -> []
-      | Some p -> [ ('i', fun () -> Image.default_portrait_filename base p) ]
+    (* TODO investigate the use of i in env *)
```
Can you explain yourself? Did you investigate it?

Halbaroth commented on lib/perso.ml:
```diff
@@ -1329,17 +1329,29 @@ let make_efam conf base ip ifam =
+    let lines = Wiki.html_of_tlsw conf s in
+    let lines =
+      (* remove enclosing <p> .. </p> if any *)
+      if List.length lines > 2 then
```
Use `List.compare_length_with lines 2 > 0` to prevent from iterating on all the list.

Halbaroth commented on lib/perso.ml:
```diff
@@ -1329,17 +1329,29 @@ let make_efam conf base ip ifam =
+      if List.length lines > 2 then
+        match lines with
+        | "<p>" :: remain ->
+            if List.hd (List.rev remain) = "</p>" then
+              List.rev (List.tl (List.rev remain))
```
I don't understand the purpose of this code? It seems that you remove closing tag `</p>` if it is present at the end of the line?

Halbaroth commented on lib/searchName.mli:
```diff
@@ -17,12 +17,9 @@ val search_by_name : config -> base -> string -> person list
 
-val search_partial_key : config -> base -> string -> person list
-(** Calls [search_key_aux] with [aux] fonction that makes calls to [search_by_name] if result is empty. *)
+val search_by_sosa : config -> base -> string -> person option
```
Please add documentation.

Halbaroth commented on lib/GWPARAM.ml:
```diff
@@ -290,6 +507,13 @@ let lang_d = ref (Legacy.lang_d : my_fun_3)
+let p_auth_sp = ref p_auth_sp
+let wrap_output = ref wrap_output
+let syslog = ref syslog
+let output_error = ref output_error
```
You can remove this code if you don't use references.

Halbaroth commented on lib/updateData.ml:
```diff
@@ -545,17 +545,28 @@ let build_list conf base =
-          Some (istr, str)
-        else None)
+        let str = sou base istr in
+        if is_name_type && str = "?" then None
```
Can you explain why "?" is a special case?

Halbaroth commented on lib/wiki.ml:
```diff
@@ -49,10 +50,11 @@ let section_level s len =
 
 (* Creates an edit button with consistent styling *)
-let make_edit_button conf fnotes ?(cnt = None) () =
+let make_edit_button conf mode fnotes ?(cnt = None) () =
```
```suggestion
let make_edit_button conf ~mode fnotes ?(cnt = None) () =
```

Halbaroth commented on lib/wiki.ml:
```diff
@@ -108,17 +110,108 @@ let str_start_with str i x =
+  let len = String.length str in
+  let rec find_index j =
+    if j >= len then None
+    else if List.mem str.[j] chars then Some j
```
`List.mem` iterates on all the list to find an element. The complexity of your implementation in the worst case is O(N * M) where N is the length of the string `str` and `M` is the length of the list `chars`. Using a hash table gives an amortized complexity in the worst case of O(N). 

```ocaml 

let find_first_char_from_list str i chars =
  let len = String.length str in
  let tbl : (char, unit) Hashtbl.t = Hashtbl.create 17 in
  List.iter (fun c -> Hashtbl.replace tbl c ()) chars;
  let rec find_index j =
    if j >= len then None
    else if Hashtbl.mem tbl str.[j] then Some j
    else find_index (j + 1)
  in
  find_index i
```

Halbaroth commented on lib/wiki.ml:
```diff
@@ -387,8 +509,8 @@ let summary_of_tlsw_lines conf short lines =
-  | Some (_, _, sfn) when conf.wizard ->
-      make_edit_button conf sfn ~cnt:(Some cnt) ()
+  | Some (_, mode, sfn) when conf.wizard ->
+      make_edit_button conf mode sfn ~cnt:(Some cnt) ()
```
```suggestion
      make_edit_button conf ~mode sfn ~cnt:(Some cnt) ()
```

Halbaroth commented on lib/util.ml:
```diff
@@ -2977,7 +3003,7 @@ let cut_words str =
     else [ String.sub str beg (i - beg) ]
   in
-  loop 0 0
+  loop 0 0 |> List.map String.trim
```
It is better to trim strings within the loop because `List.map` will iterate a second times on the list and it is not a tail recursive function. In OCaml 4, it could raise `Stack_overflow` if the list is too long.

Halbaroth commented on lib/updateDataDisplay.ml:
```diff
@@ -241,13 +229,14 @@ and eval_simple_var conf base env xx = function
       in
-      head ^ body ^ tail |> str_val
+      let tail = Printf.sprintf "<i class=\"fa fa-user fa-xs ml-1\"></i></a>" in
+      head ^ body ^ title ^ tail |> str_val
```
You can use `Format.sprintf` to concatenate `head, body, title, ...`. It is a bit more efficient and more readable.

Halbaroth commented on lib/GWPARAM.ml:
```diff
@@ -137,145 +140,359 @@ module Legacy = struct
+     && (is_semi_public p || is_related conf base p)
+     && not_private
+  || (conf.Config.public_if_no_date && access <> Private)
+  ||
```
This condition is really complex. It deserves a dedicated function with a self-documenting name.

Halbaroth commented on lib/GWPARAM.ml:
```diff
@@ -330,8 +549,17 @@ let init bname =
-  if !init_done.status && bname = !init_done.bname then ()
+  if
+    (!init_done.status && bname = !init_done.bname)
+    || not (Sys.file_exists (!bpath bname))
```
You silently ignore this error, is it what you really want?

Halbaroth commented on bin/gwd/request.ml:
```diff
@@ -104,24 +104,78 @@ let specify conf base n pl =
+      let dmy2 = Date.cdate_to_dmy_opt bi2 in
+      match dmy1, dmy2 with
+      | Some dmy1, Some dmy2 -> Date.compare_dmy dmy1 dmy2
+      | _ -> 0
```
This comparison function is probably slow. You can achieve better performances if you create a new list with extra keys on sort the list on these keys: 
```ocaml 
let l = List.rev_map (fun (p, v) -> 
      let bi = get_birth p in
      let bi =
         if bi = Date.cdate_None then get_baptism p 
         else bi 
      in
      (p, v, Date.cdate_to_dmy_opt bi))
in
List.sort (fun (_, _, k1) (_, _, k2) -> Option.compare Date.compare_dmy dmy2 dmy1) 
|> List.rev_map (fun (p, v, _) -> (p, v))
```
In the original code, you have complexity of `O(k n log n)` with `n` the length of the list `ptll` and `k` the complexity in the worst case of the huge comparison function. In the suggested code, the complexity is `O(s n log n + k n)` where `s` is the complexity of `Date.compare_dmy`. If `n` is large, `k n` becomes negligible comparing with `s n log n` and we can expect that `s` is smaller than `k`. The difference is probably small with Ancient but `get_birth` and `get_baptism` are slow if you need to read on the disk.

We could get better performances if we have a strictly increasing function from dates to integers.

Halbaroth commented on bin/gwd/request.ml:
```diff
@@ -104,24 +104,78 @@ let specify conf base n pl =
+        "MOD_NOTES" | "MOD_WIZNOTES" | "MRG_DUP_IND_Y_N" |
+        "MRG_DUP_FAM_Y_N" | "MRG_IND" | "MRG_MOD_FAM" | "MRG_MOD_IND" |
+        "MOD_DATA" | "SND_IMAGE" -> true
+      | _ -> false)
```
You can write  
```ocaml 
| Some ("ADD_FAM" | "ADD_IND" | ...) -> ...
```

Halbaroth commented on bin/gwd/request.ml:
```diff
@@ -228,6 +282,24 @@ let make_henv conf base =
   in
+  let conf =
+    if Util.p_getenv conf.env "fmode" = Some "on"
+    then { conf with henv = conf.henv @ ["fmode", Adef.encoded "on"] }
```
Adding a new variable at the end of the list is probably a mistake. The operation is slow and if the variable is already present in the environment, the new value is ignored.

Halbaroth commented on bin/gwd/request.ml:
```diff
@@ -787,6 +865,9 @@ let treat_request =
     let passwd = transl_nth conf "user/password/cancel" 1 in
+    let referer = String.split_on_char '?' (get_referer conf :> string) in
+    let referer = if List.length referer > 1 then (List.nth referer 1) else "" in
+    let referer = if referer = "" then referer else "&" ^ referer in
```
This code deserves a dedicated function.

Halbaroth commented on lib/searchName.ml:
```diff
@@ -4,6 +4,15 @@ open Config
+  case : bool;
+  exact : bool;
+  all_in : bool;
+}
```
Adding documentation is helpful to understand the meaning of each field.

Halbaroth commented on lib/searchName.ml:
```diff
@@ -4,6 +4,15 @@ open Config
+  all_in : bool;
+}
+
+(* TODO use function from Util instead? *)
```
Which function?

Halbaroth commented on lib/searchName.ml:
```diff
@@ -35,33 +44,39 @@ let select_approx_key conf base pl k =
   || (Util.is_hide_names conf p && not (Util.authorized_age conf base p))
 
 let search_by_name conf base n =
+  (* TODO use f here? why only split on the first ' '? *)
```
What is f?

Halbaroth commented on lib/searchName.ml:
```diff
@@ -35,33 +44,39 @@ let select_approx_key conf base pl k =
   let n1 = Name.abbrev (Name.lower n) in
   match String.index_opt n1 ' ' with
+  | None -> []
   | Some i ->
```
```ocaml 
match String.index n1 ' ' with 
| exception Not_found -> [] 
| i -> ...
```

Halbaroth commented on lib/searchName.ml:
```diff
@@ -123,55 +128,318 @@ type search_type =
   | PartialKey
   | DefaultSurname
 
+let match_fn_l fn_l fn1_l opts =
```
It is better to move the definition of the opts record here because `match_fn_l` is the only function that uses it.

Halbaroth commented on lib/searchName.ml:
```diff
@@ -123,55 +128,318 @@ type search_type =
+    if opts.exact then fn1 = fn2 else Mutil.contains fn2 fn1
+  in
+  let _list_equal equal fn1_l fn2_l =
+    try List.length fn1_l = List.length fn2_l && List.for_all2 equal fn1_l fn2_l
```
Use `List.equal` and you don't need to catch `Invalid_argument _` with `List.equal` ;)
As `List.equal` is not available in `OCaml < 4.12`, you should copy it from the stdlib in the module `Compat`.

Halbaroth commented on lib/searchName.ml:
```diff
@@ -123,55 +128,318 @@ type search_type =
+  | [], _, _ -> true
+  | [ fn ], _, _ when List.length fn1_l = 1 -> equal fn (List.hd fn1_l)
+  | _, true, true -> is_subsequence fn_l fn1_l
+  | _, true, false -> List.for_all (fun fn -> List.mem fn fn1_l) fn_l
```
This operation is very slow because lookup operation on linked list is linear in the size of the list. You should use Hash table or shorted array with binary search instead.

hgouraud commented:
Peut on utiliser Sys.os_type ou Sys.getenv au lieu de #ifdef ?

Halbaroth commented:
> Peut on utiliser Sys.os_type ou Sys.getenv au lieu de #ifdef ?

Oui c'est possible d'utiliser les variables définie dans `Sys` pour détecter l'OS à l'exécution.
